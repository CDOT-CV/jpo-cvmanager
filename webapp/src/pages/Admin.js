import React from 'react'
import AdminFormManager from '../components/AdminFormManager.js'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'

import '../features/adminRsuTab/Admin.css'

function Admin(props) {
  return (
    <div id="admin">
      <h2 className="adminHeader">CV Manager Admin Interface</h2>
      <Tabs>
        <TabList>
          <Tab>
            <p>RSUs</p>
          </Tab>
          <Tab>
            <p>Users</p>
          </Tab>
          <Tab>
            <p>Organizations</p>
          </Tab>
        </TabList>

        <TabPanel>
          <div className="panel-content">
            <AdminFormManager activeForm={'add_rsu'} />
          </div>
        </TabPanel>
        <TabPanel>
          <div className="panel-content">
            <AdminFormManager activeForm={'add_user'} />
          </div>
        </TabPanel>
        <TabPanel>
          <div className="panel-content">
            <AdminFormManager activeForm={'add_organization'} />
          </div>
        </TabPanel>
      </Tabs>
    </div>
  )
}

export default Admin
