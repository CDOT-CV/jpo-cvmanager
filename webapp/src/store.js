import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
