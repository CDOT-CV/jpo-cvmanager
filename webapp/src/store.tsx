import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
