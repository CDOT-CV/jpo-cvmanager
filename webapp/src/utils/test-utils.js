/**
 * This function searches for the every react-aria SSR ids in a given HTMLElement node and replace every attribute values with a static id
 *
 * This can be usefull when you're trying to generate a snapshot of components using react-aria under the hood
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceReactAriaIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: "class",
      updateFunc: (val) =>
        val.replace(
          /css-[0-9a-z]{6}-MuiTableCell-root-MuiTablePagination-root/g,
          "css-mocked-MuiTableCell-root-MuiTablePagination-root"
        ),
    },
  ];

  container.querySelectorAll("td").forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)));
      }
    });
  });
  container.querySelectorAll("input[aria-invalid]").forEach((item) => {
    delete item["aria-invalid"];
  });
  return container;
}

export { replaceChaoticIds };
