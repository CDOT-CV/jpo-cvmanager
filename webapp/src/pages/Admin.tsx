import React, { useEffect, useState } from 'react'
import AdminFormManager from '../components/AdminFormManager'
import { useDispatch } from 'react-redux'
import { updateTableData as updateRsuTableData } from '../features/adminRsuTab/adminRsuTabSlice'
import { getAvailableUsers } from '../features/adminUserTab/adminUserTabSlice'

import '../features/adminRsuTab/Admin.css'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../store'
import { Box, Tab, Tabs, Typography } from '@mui/material'
import { Link, Navigate, Route, Routes, useLocation } from 'react-router-dom'
import AdminOrganizationTab from '../features/adminOrganizationTab/AdminOrganizationTab'
import AdminRsuTab from '../features/adminRsuTab/AdminRsuTab'
import AdminUserTab from '../features/adminUserTab/AdminUserTab'
import { setRouteNotFound } from '../generalSlices/userSlice'

interface TabPanelProps {
  children?: React.ReactNode
}

function TabPanel(props: TabPanelProps) {
  const { children, ...other } = props

  return (
    <div role="tabpanel" id={`vertical-tabpanel`} aria-labelledby={`vertical-tab`} {...other}>
      <Box sx={{ p: 3 }}>
        <Typography>{children}</Typography>
      </Box>
    </div>
  )
}

function Admin() {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const location = useLocation()
  const [value, setValue] = useState(location.pathname.split('/')[location.pathname.split('/').length - 1] || 'rsus')

  const handleChange = (_e, newValue) => {
    setValue(newValue)
  }

  useEffect(() => {
    dispatch(updateRsuTableData())
    dispatch(getAvailableUsers())
  }, [dispatch])

  return (
    <div id="admin">
      <h2 className="adminHeader">CV Manager Admin Interface</h2>
      <Box
        sx={{
          flexGrow: 1,
          bgcolor: 'background',
          display: 'flex',
          width: '100%',
          height: '100%',
        }}
      >
        <Tabs
          value={value}
          onChange={handleChange}
          aria-label="Navigation"
          indicatorColor="primary"
          textColor="primary"
          orientation="vertical"
          sx={{ minWidth: 200 }}
        >
          <Tab label="RSUs" value={'rsus'} component={Link} to={'rsus'} />
          <Tab label="Users" value={'users'} component={Link} to={'users'} />
          <Tab label="Organizations" value={'organizations'} component={Link} to={'organizations'} />
        </Tabs>
        <TabPanel>
          <Routes>
            <Route index element={<Navigate to="rsus" replace />} />
            <Route path="rsus" element={<AdminRsuTab />} />
            <Route path="users" element={<AdminUserTab />} />
            <Route path="organizations" element={<AdminOrganizationTab />} />
            <Route path="*" element={() => dispatch(setRouteNotFound(true))} />
          </Routes>
        </TabPanel>
      </Box>
    </div>
  )
}

export default Admin
