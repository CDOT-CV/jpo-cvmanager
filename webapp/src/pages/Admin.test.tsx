import React from 'react'
import { render } from '@testing-library/react'
import Admin from './Admin'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import { replaceChaoticIds } from '../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'

it('should take a snapshot', () => {
  const { container } = render(
    <Provider store={setupStore({})}>
      <BrowserRouter>
        <Admin />
      </BrowserRouter>
    </Provider>
  )
  replaceChaoticIds(container)

  expect(container).toMatchSnapshot()
})
