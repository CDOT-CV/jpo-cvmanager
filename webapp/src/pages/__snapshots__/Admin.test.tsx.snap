// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div>
    <h1>
      You do not have permission to view this page
    </h1>
    <p>
      Return to public pages:
    </p>
    <a
      href="/"
    >
      Home
    </a>
  </div>
</div>
`;
